lib/tree.rb

```diff
@@ -16,45 +16,124 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n) 
+  # Space Complexity: O(log n)
+  def add(key, value = nil)
+    # raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key < current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) 
+  # Space Complexity: O(l)
   def find(key)
```

Comment:
👍 The space/time complexities are O(n) if the tree is unbalanced and O(log n) if the tree is balanced.   I'm uncertain what O(l) is?